pubsub/index.js

```diff
@@ -6,239 +6,172 @@
-    console.log(timestamp(), 'Warning: Old message', { httpapi, identify })
-  }
-
-  if (httpapi) {
-    MESSAGES.set(httpapi.insert_id, { message, payloadCount: identify ? 2 : 1 })
-    cargo.httpapi.push(httpapi)
-  }
-
-  if (identify) {
-    cargo.identify.push(identify)
-  }
-}
-
-function parseMessage (message) {
-  let { jsonPayload: event } = JSON.parse(Buffer.from(message.data, 'base64').toString())
+  const cargo = async.cargo(async (payload) => {
+    let identify_events = []
+    let http_events = payload.map(message => {
+      let event = JSON.parse(Buffer.from(message.data, 'base64').toString()).jsonPayload
+
+      if (event.Fields) {
+        event = event.Fields
+    
+        if (event.op && event.data) {
+          event = JSON.parse(event.data)
+        } else {
+          if (is.nonEmptyString(event.event_properties)) {
+            event.event_properties = JSON.parse(event.event_properties)
+          }
+    
+          if (is.nonEmptyString(event.user_properties)) {
+            event.user_properties = JSON.parse(event.user_properties)
+          }
+        }
+      }
 
-  if (event.Fields) {
-    event = event.Fields
+      if (! isEventOk(event)) {
+        logger.warn({ type: 'event.malformed', event }, 'Dropping malformed event')
+        message.ack()
+        return null
+      }
 
-    if (event.op && event.data) {
-      event = JSON.parse(event.data)
-    } else {
-      if (is.nonEmptyString(event.event_properties)) {
-        event.event_properties = JSON.parse(event.event_properties)
+      if (event.user_id) {
+        event.user_id = hash(event.user_id)
+      }
+    
+      event.insert_id = hash(event.user_id, event.device_id, event.session_id, event.event_type, event.time)
+      
+      if (IDENTIFY_VERBS.some(verb => is.assigned(event.user_properties[verb]))) {
+        identify_events.push({
+          device_id: event.device_id,
+          user_id: event.user_id,
+          user_properties: splitIdentifyPayload(event.user_properties),
+        })
       }
 
-      if (is.nonEmptyString(event.user_properties)) {
-        event.user_properties = JSON.parse(event.user_properties)
+      return event
+    }).filter(message => !!message)
+
+    // Filter out exact duplicate identify events to help prevent rate limiting
+    identify_events = identify_events.filter((outer_event, index, self) => {
+      return index === self.findIndex((inner_event) => {
+        return equal(outer_event, inner_event)
+      })
+    })
+
+    // Filter out duplicate user ids in identify events to help prevent rate limiting
+    identify_events = identify_events.filter((outer_event, index, self) => {
+      return index === self.findIndex((inner_event) => {
+        return outer_event.user_id === inner_event.user_id
+      })
+    })
+
+    if (identify_events.length > 0) {
+      let identifying_response = await retry (async bail => {
+        let identifying_response = await request('https://api.amplitude.com/identify', {
```

Comment:
This `identifying_response` shadows the outer `identifying_response`, is that expected, and will it cause any problems?